i18n/i18nTextCollector.php

```diff
@@ -44,98 +52,273 @@ class i18nTextCollector extends Object {
+	 */
 	public function setWriter($writer) {
 		$this->writer = $writer;
 	}
 
+	/**
+	 * Gets the currently assigned writer, or the default if none is specified.
+	 * 
+	 * @return i18nTextCollector_Writer
+	 */
 	public function getWriter() {
-		if(!$this->writer) $this->writer = new i18nTextCollector_Writer_RailsYaml();
+		if(!$this->writer) {
+			$this->writer = Injector::inst()->get('i18nTextCollector_Writer');
+		}
 		return $this->writer;
 	}
 	
 	/**
-	 * This is the main method to build the master string tables with the original strings.
-	 * It will search for existent modules that use the i18n feature, parse the _t() calls
-	 * and write the resultant files in the lang folder of each module.
+	 * This is the main method to build the master string tables with the
+	 * original strings. It will search for existent modules that use the
+	 * i18n feature, parse the _t() calls and write the resultant files
+	 * in the lang folder of each module.
 	 * 
 	 * @uses DataObject->collectI18nStatics()
 	 * 
 	 * @param array $restrictToModules
-	 * @param array $mergeWithExisting Merge new master strings with existing ones
-	 * already defined in language files, rather than replacing them. This can be useful
-	 * for long-term maintenance of translations across releases, because it allows
-	 * "translation backports" to older releases without removing strings these older releases
-	 * still rely on.
+	 * @param bool $mergeWithExisting Merge new master strings with existing
+	 * ones already defined in language files, rather than replacing them.
+	 * This can be useful for long-term maintenance of translations across
+	 * releases, because it allows "translation backports" to older releases
+	 * without removing strings these older releases still rely on.
 	 */	
 	public function run($restrictToModules = null, $mergeWithExisting = false) {
 		$entitiesByModule = $this->collect($restrictToModules, $mergeWithExisting);
+		if(empty($entitiesByModule)) {
+			return;
+		}
+		
 		// Write each module language file
-		if($entitiesByModule) foreach($entitiesByModule as $module => $entities) {
-			$this->getWriter()->write($entities, $this->defaultLocale, $this->baseSavePath . '/' . $module);
+		foreach($entitiesByModule as $module => $entities) {
+			// Skip empty translations
+			if(empty($entities)) {
+				continue;
+			}
+			
+			// Clean sorting prior to writing
+			ksort($entities);
+			$path = $this->baseSavePath . '/' . $module;
+			$this->getWriter()->write($entities, $this->defaultLocale, $path);
 		}
 	}
 
-	public function collect($restrictToModules = null, $mergeWithExisting = false) {
-		$modules = scandir($this->basePath);
-		$themeFolders = array();
+	/**
+	 * Gets the list of modules in this installer
+	 * 
+	 * @param string $directory Path to look in
+	 * @return array List of modules as paths relative to base
+	 */
+	protected function getModules($directory) {
+		$ignore = array('assets', 'vendor');
```

Comment:
I've removed these exemptions; We now rely on the typical module check to filter these.